### PR TITLE
Support loading gui2 themes from addons

### PR DESCRIPTION
**Pull request by babaissarkar**
The relevant gui definition should be in `<add-on-dir>/gui-theme.cfg`

**TODO**
- [x] Add a Lua function to change themes from an addon (`gui.switch_theme()`)
- [x] Make sure themes are loaded correctly just after downloading theme from addons server (that is, no restart is needed).
- [x] Allow user to switch in case of defunct theme.

### CODE REVIEW

**Comment by stevecotton** [comment]
What assumptions are you making about add-ons only having a gui subdirectory if they intend to provide a theme?

I haven't looked at the details, but _main.cfg isn't just something that's handled specially about loading add-ons, rather it's something that's handled specially [for any directory](https://wiki.wesnoth.org/PreprocessorRef#File.2Fdirectory_inclusions) by the preprocessor.

**Comment by babaissarkar** [comment]
> What assumptions are you making about add-ons only having a gui subdirectory if they intend to provide a theme?

ATM the code in `gui.cpp` (including existing code in master and this PR) operates separately from how Add-ons are loaded. If an Add-on has only a `gui/_main.cfg` file, the theme defined there would still be loaded, whether the rest of the add-on works or not.

> I haven't looked at the details, but _main.cfg isn't just something that's handled specially about loading add-ons, rather it's something that's handled specially [for any directory](https://wiki.wesnoth.org/PreprocessorRef#File.2Fdirectory_inclusions) by the preprocessor.

`gui/_main.cfg` is the entry point for the preprocessor here (once again, for both mainline `/data/gui` directory and add-on's gui directory), that is, that's where the preprocessor in instructed to start preprocessing. The writer of the add-on can write any valid preprocessor macro/wml there. If you include any directory in this special file, the preprocessor will go check that directory in the normal way, however if you don't put anything in this file nothing happens and any other files are not processed.

If you look at the code inside `gui.cpp`, it will become more clear what the gui2 subsystem is actually doing.

(Note the this PR is still in a WIP state, so details might change. What I explained is the status as of commit 2264ef1084caad2eff3613ade9cec4a607c0490e)

**Comment by soliton-** [comment]
> I actually have plans to integrate this with the add-on loading mechanism, so that the gui subsystem doesn't have to specially process a particular file, and the gui definitions will be loaded from add-ons own `_main.cfg`, but haven't looked at the specifics yet.

That does not sound like a good idea. A separate entry point is a good thing. No need for any workarounds for how addons are loaded.

**Comment by CelticMinstrel** [comment]
There's also the possibility of loading it from `<add-on>/gui.cfg`. I don't know which is better or if it makes sense to support both.

If it _is_ loading from `<add-on>/gui/_main.cfg`, it would be slightly more flexible to just try to load `<add-on>/gui`, as that permits people to omit the _main.cfg and instead provide an _initial.cfg. I don't know if there's a good reason to do that, but if it's trivial to allow it then IMO we might as well do so.

**Comment by soliton-** [comment]
I suggested `gui/_main.cfg` to mirror mainline but we can of course also choose something else.

If it seems likely that people would want to put everything in a single file (maybe they just want to customize a single dialog) then just `gui.cfg` could certainly make more sense.

**Comment by babaissarkar** [comment]
It doesn't really matter which file it is as long as there is an entry point. The UMC author can always include other files and directories from the preprocessor entry file. (That's what mainline does.)

**Comment by soliton-** [comment]
Btw, it would be cool if we could support combining multiple addon themes. So one addon could just customize a single dialog and you can install multiple such addons and combine them. (If they would have overlapping customizations then I think it'd be ok if it's simply unspecified which one is used.)

**Comment by stevecotton** [comment]
I think we should (already do?) reserve filenames beginning with an underscore for special handling. So if this has to be done with a filename rather than a WML tag, I suggest `_gui.cfg`.

**Comment by soliton-** [comment]
Reserving files at the top level, beginning with an underscore sounds like a good idea. The existing `achievements.cfg` and `cores.cfg` do not fit that though. Not sure if there is more.

**Comment by CelticMinstrel** [comment]
> I think we should (already do?) reserve filenames beginning with an underscore for special handling. So if this has to be done with a filename rather than a WML tag, I suggest `_gui.cfg`.

Given that we already have reserved toplevel files, I don't think there's much point in adding a new convention now for adding an underscore prefix.

**Comment by stevecotton** [comment]
I'd prefer to deprecate achievements.cfg and cores.cfg in 1.20, in preference to versions of those names with an underscore prefix.

**Comment by CelticMinstrel** [comment]
I'd definitely prefer _not_ to do that. The underscore-prefixed files are special handling in the preprocessor, but `cores.cfg` and `achievements.cfg` are special handling in the engine. They are special in different ways.

**Comment by CelticMinstrel** [comment]
Also, I just realized that there's a problem with loading the `gui` directory as the GUI2 configuration for an add-on. We've started using a `gui` directory to contain custom dialogs implemented in Lua, so attempting to load the `gui` directory will fail _right now_ in Secrets of the Ancients, for example. (Though this assumes mainline campaigns are loaded as add-ons, which I think they aren't currently.)

It would be easy to fix (just add a blank `_main.cfg` in that folder), but it could also be dodged by having the file actually loaded be a specifically-named file (like `<addon>/gui.cfg` or `<addon>/hardwired/gui.cfg` or whatever else we decide on) which just conventionally has the content `{../gui}` in mainline content.

**Comment by stevecotton** [comment]
> The underscore-prefixed files are special handling in the preprocessor, but `cores.cfg` and `achievements.cfg` are special handling in the engine. They are special in different ways.

While it's a big separation, that's still an implementation detail of "your add-on can run on the Wesnoth platform". When the target audience is UMC authors, having a reserved prefix for "these names are reserved for the Wesnoth platform" seems clearer than going in to the detail of what's handled by which part of Wesnoth's implementation.

**Comment by babaissarkar** [comment]
What are the thoughts on making the gui2 theme loadable via wml tag/lua instead of relying on special filenames?
I'd also prefer to name the entry point file `gui-theme.cfg` instead of plain `gui.cfg`, it might cause confusion with custom dialogs added by addons as said earlier here.

**Comment by soliton-** [comment]
> What are the thoughts on making the gui2 theme loadable via wml tag/lua instead of relying on special filenames?

How would that work? In general combining if an addon is loaded or not with if a theme is enabled or not does not really make sense unless the theme should specifically just be active when you play the addon content. In general I expect addon content to rather want to switch the "in-game" theme rather than the general GUI2 theme but that's a different topic really.

I agree that the filename discussion should not block this PR.

**Comment by stevecotton** [comment]
We already have the concept that the `[campaign]` tags are loaded for the main menu, and contain information that results in loading other parts of an add-on. We could have a similar mechanism for the part of a theme add-on that needs to be accessible when the theme isn't in use, the part that's used in a menu for enabling/disabling that theme.

**Comment by babaissarkar** [comment]
Changed the entry point to `gui-theme.cfg`. This makes its intent clear (cannot be confused with a dialog file or a schema file). (There's a `gui.cfg` inside `/data/schema`. Although there's not much chance of filename conflict, I'd prefer not to use the same name for an entirely different purpose.)

File: src/gui/gui.cpp [hunk 1]
```diff
@@ -51,10 +51,32 @@ void init()
...
+		for(const std::string& umc : user_dirs) {
+			const std::string gui_file = umc + "/gui-theme.cfg";
+			if(filesystem::file_exists(gui_file)) {
+				config gui_cfg;
+				cache.get_config(gui_file, gui_cfg);
+				validator.validate(gui_cfg, "gui", 0, gui_file);
+				cfg.append(gui_cfg);
+			}
+		}
+
```
Comment (soliton-) [inline src/gui/gui.cpp:79]: Comment and error don’t fit anymore.

Comment (babaissarkar) [reply src/gui/gui.cpp:83]: are the changes in 3c074a027c21b434e8834388e065779ac5903a1a adequate?

File: src/gui/gui.cpp [hunk 2]
```diff
@@ -66,22 +70,22 @@ void init()
...
 			}
 		}
 
 	} catch(const config::error& e) {
 		ERR_GUI_P << e.what();
-		ERR_GUI_P << "Setting: could not read file 'data/gui/_main.cfg'.";
+		ERR_GUI_P << "Setting: could not read gui file: " << current_file;
 	} catch(const abstract_validator::error& e) {
-		ERR_GUI_P << "Setting: could not read file 'data/schema/gui.cfg'.";
+		ERR_GUI_P << "Setting: could not read schema file: " << schema_file;
```
Comment (CelticMinstrel) [inline src/gui/gui.cpp:87]: Though I don't have a preference about this change, it's technically not necessary, is it? Since all these files do use the same schema. It's not like it can hurt though…

File: src/gui/gui.cpp (hunk 1 above)
Comment (soliton-) [reply src/gui/gui.cpp:82]: Using `config_cache` seemed like a good idea.

**Comment by Pentarctagon** [comment]
Is there anything else left to do for this PR aside from resolving the conflict?

**Comment by babaissarkar** [comment]
> Is there anything else left to do for this PR aside from resolving the conflict?

Added a task list above.

**Comment by CelticMinstrel** [comment]
Please add documentation for the new function in utils/emmylua/gui.lua

Note: It's not actually EmmyLua markup anymore but rather the markup supported by the VSCode Lua plugin, which is based on EmmyLua but adds several extensions.

**Comment by Vultraz** [comment]
For the record, I'm opposed to adding this. I don't think we should be exposing this functionality to players yet, given we have yet to port the in-game HUD to GUI2 (if we ever actually do), and given that that's the best usecase we have for this feature, exposing theming of the entire UI feels to me like overkill.

If people think it's worth having, I won't stand in the way of merging, but I think it should at the very least be held off for now.

**Comment by soliton-** [comment]
This is pretty much entirely unrelated to the in-game theme. Depending on how the in-game theme might look in the future there could be some overlap but this is specifically to provide a theme potentially for the entire UI. I don't see why this would depend on possible future in-game theme changes.

What would be good to figure out is what's the worst that can happen with a bad UI theme. Ideally there would be a way to reset to default at startup if the current theme is broken in some way without having to edit the preference file. For example it might be easy to always take the titlescreen and preferences dialog from the default theme. Then at a minimum you can get to the preferences at start and change the theme if there is an issue. Maybe there are less restrictive approaches as well.

File: src/wesnoth.cpp [hunk 3]
```diff
@@ -898,6 +898,8 @@ static int do_gameloop(commandline_options& cmdline_opts)
 		case gui2::dialogs::title_screen::RELOAD_GAME_DATA:
 			gui2::dialogs::loading_screen::display([&config_manager]() {
 				config_manager.reload_changed_game_config();
+				gui2::init();
```
Comment (soliton-) [inline src/wesnoth.cpp:901]: I suppose this happens on F5 and addon install? In the addon install case it doesn't really make sense to parse mainline themes again. Don't know if you can currently decide which case it is though and it might not make much of a difference..

File: src/gui/gui.cpp [hunk 4]
```diff
@@ -42,17 +42,43 @@ void init()
...
+				user_campaign_dir, &user_files, &user_dirs, filesystem::name_mode::ENTIRE_FILE_PATH);
+		}
+
+		for(const std::string& umc : user_dirs) {
+			const std::string gui_file = umc + "/gui-theme.cfg";
+			current_file = filesystem::get_short_wml_path(gui_file);
+			if(filesystem::file_exists(gui_file)) {
+				config addon_cfg;
+				schema_validation::schema_validator validator(filesystem::get_wml_location(schema_file).value());
+				read(addon_cfg, *preprocess_file(gui_file, &preproc), &validator);
```
Comment (soliton-) [inline src/gui/gui.cpp:72]: Would be good if each error is handled separately so one broken addon does not disable all other addon themes.

File: src/game_initialization/playcampaign.cpp [hunk 5]
```diff
@@ -205,6 +206,7 @@ level_result::type campaign_controller::play_game()
 				}
 			}
 
+			gui2::switch_theme(prefs::get().gui2_theme());
```
Comment (soliton-) [inline src/game_initialization/playcampaign.cpp:209]: This could use a comment. If this is supposed to change the theme back when some scenario changed it then I don't think this is the right place.

File: src/serialization/schema_validator.cpp [hunk 6]
```diff
@@ -432,10 +436,12 @@ void schema_validator::validate(const config& cfg, const std::string& name, int
...
-	// clear cache
-	auto cache_it = cache_.top().find(&cfg);
-	if(cache_it != cache_.top().end()) {
-		cache_it->second.clear();
+	if (!cache_.empty()) {
+		// clear cache
+		auto cache_it = cache_.top().find(&cfg);
+		if(cache_it != cache_.top().end()) {
+			cache_it->second.clear();
+		}
```
Comment (soliton-) [inline src/serialization/schema_validator.cpp:444]: These changes are there to support re-using a validator?

File: src/hotkey/command_executor.cpp [hunk 7]
```diff
@@ -361,6 +362,7 @@ bool command_executor::do_execute_command(const hotkey::ui_command& cmd, bool pr
 			quit_confirmation::quit_to_desktop();
 			break;
 		case HOTKEY_QUIT_GAME:
+			gui2::switch_theme(prefs::get().gui2_theme());
```
Comment (soliton-) [inline src/hotkey/command_executor.cpp:365]: I guess this is also there to change the theme back when some scenario changed it?

**Comment by CelticMinstrel** [comment]
> For example it might be easy to always take the titlescreen and preferences dialog from the default theme.

Though this certainly makes some amount of sense, the titlescreen is also one of the things someone would be most likely to want to change with a custom theme… and is the _only_ thing changed by the currently-existing alternate theme (though that's not an add-on theme).

**Comment by soliton-** [comment]
I don't think that's the worst that can happen. Pretty sure you could make a theme with a titlescreen and no buttons for example. So a less restrictive solution could be to check that the titlescreen contains a visible button to get to the preferences and preferences is always taken from the default theme.

**Comment by Pentarctagon** [comment]
I doubt such a theme would work? Unless the titlescreen is already coded to assume any or all its buttons could be missing, which I don't see any reason for it to do.

**Comment by CelticMinstrel** [comment]
> Pretty sure you could make a theme with a titlescreen and no buttons for example.

Pretty sure that would cause Wesnoth to crash because the buttons are mandatory.

> preferences is always taken from the default theme.

This I think I can agree on.

File: src/serialization/schema_validator.cpp (hunk 6 above)
Comment (babaissarkar) [reply src/serialization/schema_validator.cpp:444]: It stops wesnoth from crashing. When testing addon themes I noticed in some cases this can be called when `cache_` is empty, which causes segfault.

File: src/hotkey/command_executor.cpp (hunk 7 above)
Comment (babaissarkar) [reply src/hotkey/command_executor.cpp:365]: Yes.

File: src/gui/gui.cpp [hunk 8]
```diff
@@ -62,10 +88,15 @@ void init()
...
 		const std::string id = g["id"];
 
-		auto iter = guis.emplace(id, gui_definition(g)).first;
+		auto [iter, is_unique] = guis.emplace(id, gui_definition(g));
 
-		if(id == "default") {
-			default_gui = iter;
+		if (!is_unique) {
+			ERR_GUI_P << "GUI Theme ID '" << id << "' already exists.";
+			current_gui = default_gui;
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/gui/gui.cpp:95]: In case there are two themes with same ID, it's supposed to use default theme instead for the second one. In other words, the first one takes precedence.

File: src/wesnoth.cpp (hunk 3 above)
Comment (babaissarkar) [reply src/wesnoth.cpp:901]: > I suppose this happens on F5 and addon install
Yes.

**Comment by soliton-** [comment]
Well, not no buttons but not reachable buttons then? If it's really trivially determinable when you can reach preferences that's cool. I just don't expect it to be that easy.

File: src/gui/gui.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (soliton-) [reply src/gui/gui.cpp:95]: I don't get it. What does "use default theme instead for the second one" mean? The duplicate is not inserted at all. Also this loop isn't for selecting a theme is it? That's `switch_theme` below.

[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/gui/gui.cpp:95]: > The duplicate is not inserted at all.

Didn't notice that. Will removing `current_gui = default_gui;` be enough?

File: src/gui/gui.cpp [hunk 9]
```diff
@@ -62,10 +88,15 @@ void init()
...
+		} else {
+			if(id == "default") {
```
Comment (babaissarkar) [inline src/gui/gui.cpp:97]: I'll need to check if the three lines here make any sense either. A duplicate "default" won't be inserted anyway.

File: src/gui/gui.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (soliton-) [reply src/gui/gui.cpp:95]: Yes, that is the part that makes no sense to me.

File: src/gui/gui.cpp (hunk 9 above)
Comment (soliton-) [reply src/gui/gui.cpp:97]: I think it's correct though IMO even better would be to only check mainline themes for this.

Comment (babaissarkar) [reply src/gui/gui.cpp:108]: > IMO even better would be to only check mainline themes for this.

Done.

File: src/game_initialization/playcampaign.cpp (hunk 5 above)
Comment (babaissarkar) [reply src/game_initialization/playcampaign.cpp:211]: > This could use a comment.

Added.

> If this is supposed to change the theme back when some scenario changed it then I don't think this is the right place.

What would be a better place?

File: src/gui/gui.cpp (hunk 4 above)
Comment (babaissarkar) [reply src/gui/gui.cpp:85]: Done.

File: src/gui/gui.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/gui/gui.cpp]: Removed.

File: src/game_initialization/playcampaign.cpp (hunk 5 above)
Comment (soliton-) [reply src/game_initialization/playcampaign.cpp:211]: The `campaign_controller` destructor perhaps.

Comment (babaissarkar) [reply src/game_initialization/playcampaign.cpp]: Moved to `campaign_controlle`r's destructor.

**Comment by babaissarkar** [comment]
Changes in https://github.com/wesnoth/wesnoth/pull/9057/commits/72564e0328e954dc530c99963669b00b0b2d24b6

Apparent there was already a lot of validation going on while loading a theme. What was actually crashing wesnoth was the uncaught `wml_exception`. Catching this exception causes the fallback to default theme to work correctly. I've added some error messages and improved other relevant ones to give the user more idea of what's exactly failing their theme.
In short: it's not very easy to make a defunct theme that passes so many layers of validation checks and still makes wesnoth non-functional.

**Comment by babaissarkar** [comment]
The titlescreen buttons issue still needs to be dealt with, as it uses 4 arg version of `find_widget` which allows the titlescreen to appear even with missing buttons.

**Comment by CelticMinstrel** [comment]
> Well, not no buttons but not reachable buttons then? If it's really trivially determinable when you can reach preferences that's cool. I just don't expect it to be that easy.

Ah, putting the buttons in a place were they can't be clicked? I suppose that might be possible. Not sure though.

**Comment by babaissarkar** [comment]
> So a less restrictive solution could be to check that the titlescreen contains a visible button to get to the preferences and preferences is always taken from the default theme.

I did this BTW.

File: src/config.hpp [hunk 10]
```diff
@@ -634,6 +634,13 @@ class config
 	const config& find_mandatory_child(config_key_type key, const std::string &name,
 		const std::string &value) const;
 
+	optional_config_impl<config> find_child_recursive(config_key_type key, const std::string &name,
```
Comment (CelticMinstrel) [inline src/config.hpp:637]: Why do you need a non-const version?

**Comment by CelticMinstrel** [comment]
> > So a less restrictive solution could be to check that the titlescreen contains a visible button to get to the preferences and preferences is always taken from the default theme.
> 
> I did this BTW.

Your check verifies the item exists but not that it is visible.

File: src/config.hpp (hunk 10 above)
Comment (babaissarkar) [reply src/config.hpp:637]: Was just following the convention in that file, but I technically don't need the non-const version.

**Comment by babaissarkar** [comment]
> Your check verifies the button exists but not that it is visible.
> 
> I'm not sure it's generically possible to verify it's visible without first building the widget.

Are you referring to [gui2::widget::visibility](https://devdocs.wesnoth.org/classgui2_1_1widget.html#ab1156d4f673335503574518caf508873), or something else?

File: src/config.hpp (hunk 10 above)
Comment (CelticMinstrel) [reply src/config.hpp:637]: I double-checked the implementation of `optional_config_impl`, and it looks like it _is_ an optional reference, so that probably is the reason to have two.

**Comment by CelticMinstrel** [comment]
> > Your check verifies the button exists but not that it is visible.
> > I'm not sure it's generically possible to verify it's visible without first building the widget.
> 
> Are you referring to [gui2::widget::visibility](https://devdocs.wesnoth.org/classgui2_1_1widget.html#ab1156d4f673335503574518caf508873), or something else?

I don't think that can be controlled from the WML, can it? I was imagining someone hiding it by squirrelling it away on page 3 of a `multi_page` that the preferences dialog doesn't expect and thus provides no means to navigate.

File: src/config.hpp (hunk 10 above)
Comment (Vultraz) [reply src/config.hpp:637]: If we can use one, use one. Using `const_cast` is really not optimal, because there's no guarantee that the const version won't have side effects (which it shouldn't).

File: src/gui/gui.cpp [hunk 11]
```diff
@@ -38,34 +38,94 @@ void init()
...
+	}
+
+	//
+	// Parse GUI definitions from addons
+	//
+	for(const config& g : addons_cfg.child_range("gui")) {
+		const std::string id = g["id"];
+
+		try {
+			auto [iter, is_unique] = guis.emplace(id, gui_definition(g));
```
Comment (Vultraz) [inline src/gui/gui.cpp:121]: Use 
```cpp
auto [iter, is_unique] = guis.try_emplace(id, g);
```
to avoid creating unnecessary `gui_definition` objects.

File: src/config.hpp (hunk 10 above)
Comment (CelticMinstrel) [reply src/config.hpp:637]: But if we want to return a reference, we probably need both.

Comment (babaissarkar) [reply src/config.hpp:637]: > Using `const_cast` is really not optimal,

Will removing the `const_cast` but keeping that function work?
By the same logic, any other functions using `const_cast` should be adjusted as well.

**Comment by babaissarkar** [comment]
> I was imagining someone hiding it by squirrelling it away on page 3 of a multi_page that the preferences dialog doesn't expect and thus provides no means to navigate.

I can probably blacklist stacked_widget/multi_page from being used in title_screen.
But the bigger issue is that there are probably other ways to make widgets invisible. Wouldn't it be better to leave the more pathological cases to user discretion and deal with them on a case by case basis?
Alternative, title_screen could be blacklisted and a limited subset of its appearance parameters could be exposed in `gui_theme.cfg` (for example, background or sidebar positioning)

**Comment by babaissarkar** [comment]
> That would allow people to _legitimately_ do something like the above _and_ provide the means to navigate it right in the WML.

`tab_container` at least allows moving between tabs, which could be themed to look a bit different. However, stacked_widget/multi_page are reliant on code to switch their layers/pages.

File: src/gui/gui.cpp (hunk 11 above)
Comment (babaissarkar) [reply src/gui/gui.cpp]: Updated.

File: src/config.hpp (hunk 10 above)
Comment (babaissarkar) [reply src/config.hpp]: Rewrote `find_child_recursive` to not use `const_cast`.

Comment (CelticMinstrel) [reply src/config.hpp]: Personally I think const_cast is a perfectly legitimate way to handle these kinds of function pairs, where one returns a const reference and the other returns a non-const reference, and the latter uses const_cast. I'd certainly prefer that to writing the entire function twice.

**Comment by CelticMinstrel** [comment]
> > I was imagining someone hiding it by squirrelling it away on page 3 of a multi_page that the preferences dialog doesn't expect and thus provides no means to navigate.
> 
> I can probably blacklist stacked_widget/multi_page from being used in title_screen. But the bigger issue is that there are probably other ways to make widgets invisible. Wouldn't it be better to leave the more pathological cases to user discretion and deal with them on a case by case basis?

Yeah, making it perfect is probably hard or nearly impossible. I guess it doesn't need to be perfect. Blacklisting multi_page and stacked_widget might be good enough? Though, I'll note that stacked_widget _might_ not result in it being invisible, as it shows every layer at the same time… but even if it's not _invisible_ I bet it would make it unclickable.

> Alternative, title_screen could be blacklisted and a limited subset of its appearance parameters could be exposed in `gui_theme.cfg` (for example, background or sidebar positioning)

I don't think I like this idea. It seems _too_ limiting.

File: src/config.hpp (hunk 10 above)
Comment (Pentarctagon) [reply src/config.hpp]: It could probably be addressed with a macro or maybe templates? const_cast is something to avoid if at all possible just generally is my understanding.

Comment (Pentarctagon) [reply src/config.hpp]: Also yeah, could just remove the non-const version.

Comment (CelticMinstrel) [reply src/config.hpp]: I think this kind of situation is _precisely_ one of the cases that are an exception to that rule.

**Comment by Vultraz** [comment]
This PR has gotten bikeshed enough. Let's drop the blacklisting (so, revert the changes to `gui_definition.cpp`; we can handle disabling customization on a per-window basis later) and the recursive config functions and I'll merge this by tomorrow.

**Comment by babaissarkar** [comment]
> This PR has gotten bikeshed enough. Let's drop the blacklisting (so, revert the changes to `gui_definition.cpp`; we can handle disabling customization on a per-window basis later) and the recursive config functions and I'll merge this by tomorrow.

Done. I've also squashed and rebased the commits.
